### ISSUE TITLE
fluent validation for jsonPatchDocument

### ISSUE THREAD

**Issue by arielcloud** [JeremySkinner/FluentValidation#308]
Is there any way to fluent-validate jsonPatchDocument before applying it?
for example, if I want to disable the client to set the ID, I have the rule: `RuleFor(a => a.AccountId).Null();`, now, I want that he also wont edit the id, or set it to null, so I want to check before applying it that no 

`[
    {
      "value": null,
      "path": "/AccountId",
      "op": "replace"
    }
  ]`

operation exists. after I apply it, the validation will passed since the value is null, and that's legal, but it's not really legal to edit it by the client...

thanks!

**Comment by JeremySkinner** [comment]
Not really...FluentValidation designed to validate the properties of objects after they've been set (eg, for a viewmodel/input model scenario). If you need to pre-validate the incoming JSON then you'd really need to do that outside of FluentValidation.

**Comment by arielcloud** [comment]
Would not it be great to add it as a future enhancement? 
since jsonPatchDocument is now part of asp.net core, and it's going to be in wide use over time?
since all my validations sit in FV files it really was great if I can validate Patch using this rules.

I understand that you see it as a separate part which should be implemented elsewhere. thanks anyway.

**Comment by JeremySkinner** [comment]
What would you imagine support for this would look like in FV?

**Comment by arielcloud** [comment]
I would imagine `ApplyToAndThrow(T patched, IValidator<T> tValidator)` function that I can use instead of  `ApplyTo(T patched)`, which would apply just if all rules saved:
```
private readonly IValidator<Account> _accountValidator;
public AccountDTO Patch(long id, [FromBody]JsonPatchDocument<Account> patch)
        {
            var account = GetAccount(id);
            patch.ApplyToAndThrow(account, _accountValidator);
```

**Comment by JeremySkinner** [comment]
The problem is how would this be implemented internally? FV works by reading the property values off the object being validated- at this stage the values haven't actually been set yet so I can see no way that this could work.

**Comment by arielcloud** [comment]
I implemented this by applying it to x, then validateAndThrow(x) before `context.saveChanges()`. so, if not valid - it wont save to DB. 
regarding the edge case I described above (of replace to null), I looped on all operations in patch.operations, and checked for each FV `property.Null()` rule - that no operation.path lead to this property.

maybe there are another edge cases I still didn't thought of. but I believe there are ways that this could work.
